File: Code/Handout/dynamic/womandsimulation.py
# Author: Jorge L. García
# Date: 7/1/2014

#Import packages
import numpy as np
from scipy.stats import norm

#Set seed
np.random.seed(0)

#Model Parameters
N = 1000; T = 6; betak = .5; betan = .2; sigmaeps = 0.4; pi = .2; 
gamma1 = .8; sigmaeta = 1.0; covepseta = .3; gamma2 = .9; delta = .85;
covepsxi = covepseta - sigmaeps**2
covetaxi = sigmaeta**2 - covepseta
sigmaxi = np.sqrt(sigmaeta**2 + sigmaeps**2 - 2*covepseta)

#Simimulation Parameters
y_lb = 0.0; y_up = 10.0;
k_lb = 0.0; k_up = 5.0;
z_lb = 0.0; z_up = 5.0;
n_lb = 0; n_up = 3

#Simulate observed variables
#y
y = np.random.uniform(low=y_lb,high=y_up,size=(N,T))

#z,k,n
z = np.zeros((N,T))
kappa = np.zeros((N,T))
n = np.zeros((N,T))

z_N = np.random.uniform(low=z_lb,high=z_up,size=N)
kappa_N = np.random.uniform(low=k_lb,high=k_up,size=N)
n_N = np.random.random_integers(low=n_lb,high=n_up,size=N)

for t in range(T):
    z[:,t] = z_N
    kappa[:,t] = kappa_N
    n[:,t] = n_N

#Simulate unobserved variables
epseta = np.random.multivariate_normal([0,0],[[sigmaeta**2, covepseta],[covepseta, sigmaeps**2]],(N,T))
eta = epseta[:,:,0]
eps = epseta[:,:,1]

#Construct latent and observed decision
#Non-stochastic components of the utility function
def W1(z,h,n):
    return gamma1*z + gamma2*h - pi*n
    
def W0(kappa,n):
    return betak*kappa + betan*n

#xi_star @T
def xi_starTm0(z,h,n,kappa):
    return W1(z,h,n) - W0(kappa,n)

#Emax @T
def EmaxTm0(z,h,n,kappa):
    cdf1 = norm(0,1).cdf(xi_starTm0(z,h,n,kappa)/sigmaxi)
    pdf  = norm(0,1).pdf(-xi_starTm0(z,h,n,kappa)/sigmaxi)
    cdf0 = norm(0,1).cdf(-xi_starTm0(z,h,n,kappa)/sigmaxi)
    return W1(z,h,n)*cdf1 + W0(kappa,n)*cdf0 + sigmaxi*pdf
    
#xi_star @ 1, ..., T-1
for t in range(1,T):
    exec("def xi_starTm" + str(t) + "(z,h,n,kappa):\
    return W1(z,h,n) - W0(kappa,n)   \
    + delta*EmaxTm" + str(t-1) + "(z,h+1.0,n,kappa)\
    - delta*EmaxTm" + str(t-1) + "(z,h,n,kappa)")

#Emax @ t = 2,...,T-1 
for t in range(1,T-1):
    exec("def EmaxTm" + str(t) + "(z,h,n,kappa):\
    cdf1 = norm(0,1).cdf(xi_starTm" + str(t) + "(z,h,n,kappa)/sigmaxi);\
    pdf = norm(0,1).pdf(-xi_starTm" + str(t) + "(z,h,n,kappa)/sigmaxi);\
    cdf0 = norm(0,1).cdf(-xi_starTm" + str(t) + "(z,h,n,kappa)/sigmaxi);\
    return (W1(z,h,n) + delta*EmaxTm" + str(t-1) + "(z,h+1.0,n,kappa))*cdf1 + \
    (W0(kappa,n) + delta*EmaxTm" + str(t-1) + "(z,h,n,kappa))*cdf0 + \
    sigmaxi*pdf")

xi_star = np.zeros((N,T)); d = np.zeros((N,T)); h = np.zeros((N,T))

for t in range(0,T):
    exec("xi_star[:," + str(t) + "] = xi_starTm" + str(T-1-t) + "(z[:," + str(t) + "],\
    h[:," + str(t) + "], n[:," + str(t) + "], kappa[:," + str(t) + "])")
    d[:,t] = eta[:,t] - eps[:,t] > -xi_star[:,t]    
    if t < T-1:   
        h[:,t+1] = h[:,t] + d[:,t]
               
#wage
w = d*(gamma1*z + gamma2*h + eta)

np.savetxt('womanddata', np.column_stack((y, z, kappa, n, d, w, h)), fmt= '%8.3f')

File: Code/Handout/dynamic/womandsimulation.jl
# Author: Jorge L. García
# Date: 7/1/2014

# Import packages
using Distributions
using Optim
using DataFrames

# Set a seed
srand(1)

# Parameters
# model
N = 1000; T = 6; betak = .5; betan = .2; sigmaeps = .4; pi = .2;
gamma1 = .8; sigmaeta = 1; covepseta = .3; gamma2 = .9; delta = .85;
covepsxi = covepseta - sigmaeps^2
covetaxi = sigmaeta^2 - covepseta
sigmaxi  = sqrt(sigmaeta^2 + sigmaeps^2 - 2*covepseta)

# simulation
y_lb = 0; y_ub = 10
z_lb = 0; z_ub = 5
k_lb = 0; k_ub = 5
n_lb = 0; n_ub = 3

# Simulate
# y
y = zeros(N,T)
for t = 1:T
        y[:,t] = [rand(Uniform(y_lb,y_ub)) for i = 1:N]
end

# z,kappa,n (static over time)
z = zeros(N,T)
kappa = zeros(N,T)
n = zeros(N,T)

z_N = [rand(Uniform(z_lb,z_ub)) for i = 1:N]
kappa_N = [rand(Uniform(k_lb,k_ub)) for i = 1:N]
n_N = [rand(Uniform(n_lb,n_ub)) for i = 1:N]

# z,kappa,n (wide)
for t = 1:T
        z[:,t] = z_N
        kappa[:,t] = kappa_N
        n[:,t] = n_N
end

# unobserved
eps = zeros(N,T)
eta = zeros(N,T)

for t = 1:T
        epseta =  [rand(MvNormal([sigmaeps^2 covepseta; covepseta sigmaeta^2])) for i = 1:N]
        for i = 1:N
                eps[i,t] = epseta[i,1][1]
                eta[i,t] = epseta[i,1][2]
        end
end

# non-stochastic terms of utility
function W1(z,h,n)
        return gamma1*z + gamma2*h - pi*n
end

function W0(kappa,n)
        return betak*kappa + betan*n
end

# xistar, t = T - 0
function xistarTm0(z,h,n,kappa)
        return W1(z,h,n) - W0(kappa,n)
end

# Emax, t = T - 0
function EmaxTm0(z,h,n,kappa)
        cdf1 = cdf(Normal(0,1), xistarTm0(z,h,n,kappa)/sigmaxi)
        pdf0 = pdf(Normal(0,1), -xistarTm0(z,h,n,kappa)/sigmaxi)
        cdf0 = cdf(Normal(0,1), -xistarTm0(z,h,n,kappa)/sigmaxi)
        return W1(z,h,n).*cdf1 + W0(kappa,n).*cdf0 + sigmaxi*pdf0
end

# xistar, t = T - 1,..., 1
for t = 1:T-1
        eval(parse("function xistarTm" * string(t) * "(z,h,n,kappa)"
        * "W1(z,h,n) - W0(kappa,n)"
        * "+" * "delta*EmaxTm" * string(t-1) * "(z,h+1,n,kappa)" * "-" * "delta*EmaxTm" * string(t-1) * "(z,h,n,kappa)"
        * "end"))
end

# EmaxTmt,   t = T -1, ..., 2
for t = 1:T-1
        eval(parse("function EmaxTm" * string(t) * "(z,h,n,kappa)"
        * "W1(z,h,n) - W0(kappa,n)"
        * "+" * "delta*EmaxTm" * string(t-1) * "(z,h+1,n,kappa)" * ".*"
        * "cdf(Normal(0,1), xistarTm" * string(t) * "(z,h,n,kappa)/sigmaxi)"
        * "+" * "delta*EmaxTm" * string(t-1) * "(z,h,n,kappa)" * ".*"
        * "cdf(Normal(0,1), -xistarTm" * string(t) * "(z,h,n,kappa)/sigmaxi)"
        * "+" * "sigmaxi*pdf(Normal(0,1), -xistarTm" * string(t) * "(z,h,n,kappa)/sigmaxi)"
        * "end"))
end

#xistar vector function
xistarvf = [xistarTm0]
for t = 1:T-1
        xistarvf = [eval(parse("xistarTm" * string(t)));xistarvf]
end

# xistar, d, h
xistar = zeros(N,T); d = zeros(N,T); h = zeros(N,T)

for t = 1:T
        xistar[:,t] = xistarvf[t](z[:,t], h[:,t], n[:,t], kappa[:,t])
        for i = 1:N
                d[i,t] = eta[i,t] - eps[i,t] + xistar[i,t] > 0
        end
        if t < T
                h[:,t+1] = h[:,t] + d[:,t]
        end
end

# w
w = d.*(gamma1*z + gamma2*h + eta)

# Save to a .csv file
womandsample = [y z kappa n d w h]
womandsample = DataFrame(womandsample)
writetable("womandsample.csv",womandsample,header=true)
